### PR TITLE
Static groups UI

### LINKED ISSUE #5472: Static Groups
As a Network Automation Engineer, I want the ability to create and manage static groups that function much like dynamic groups, except their membership is statically managed.

This is to support grouping use cases in which dynamic membership is problematic or otherwise unwanted. These static groups will also be used to underpin a refactor of the way dynamic groups operate, to solve critical performance issues.

AC:
- Groups are defined per content type
- Membership is based upon being explicitly related to the group
- **Not** implemented as a TreeModel to start

### PR DESCRIPTION
**Pull request by glennmatthews**
# Closes #5472
# What's Changed

Continuation of the work from #5631.

- Automatically add a `static_groups` filter to any subclass of BaseFilterSet that is static-group-associable
  - This ended up looking similar to the logic that we had in `get_filterset_parameter_form_field()` for tag, status, and role generic filters, but I liked my implementation better as it's on the filterset itself rather than a separate function. So I refactored how those three filters are set up to similarly move their logic to the filterset or mixin and out of the function.
- Add "Update Static Groups" button to object_list.html generic template
  - This opens a simple modal rather than sending the user to a separate view, similar to the Saved View modal in #5639.
  - Add custom form and custom view for bulk-assigning selected objects to static groups as backing for this modal.
- Add "Static Groups" tab to object_retrieve.html generic template (and device.html template since that overrides it)
  - Like the "Contacts" tab, this is implemented as part of the base view/template rather than a separate standalone view.
- Add `tenant` field to the StaticGroup model, and update factory, filterset, and tests accordingly
  - [x] TODO might need some model documentation update for this
- Add full UI for StaticGroups, including nav menu entry
- Add limited subset of UI for StaticGroupAssociations
- Add static groups count to Tenant detail view and re-sort its contents alphabetically.

# Screenshots

image
image
image
image
image
image
image
image
image

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- n/a Example App Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

### CODE REVIEW

**Comment by glennmatthews** [comment]
Added static groups field (like tags field) to object edit forms:

image

It'd be good to add a generic test for this, but we don't even have good generic test coverage for the tags field yet.

File: changes/5472.added [hunk 1]
```diff
@@ -1 +1,3 @@
 Added `StaticGroup` and `StaticGroupAssociation` data models, used for statically defining groups of Nautobot objects.
+Added `feature` filter to `/api/extras/content-types/`.
+Added `static_group` filter to all applicable models via the `BaseFilterSet` class.
```
Comment (glennmatthews) [inline changes/5472.added:3]:
```suggestion
Added `static_group` filter to all applicable models via the `BaseFilterSet` class.
Added `static_group` to applicable model create/edit forms via the `StaticGroupModelFormMixin` class (included in `NautobotModelForm` class automatically).
```

File: nautobot/core/filters.py [hunk 2]
```diff
@@ -716,8 +715,23 @@ def get_filters(cls):
         """
         filters = super().get_filters()
 
+        if "static_groups" not in filters and getattr(cls._meta.model, "is_static_group_associable_model", False):
```
Comment (glennmatthews) [inline nautobot/core/filters.py:718]: Might make sense to move this out of `BaseFilterSet` and over to a mixin included in `NautobotModelFilterSet` instead, just for self-consistency's sake? Though we have a lot of filtersets that are derived from BaseFilterSet only, especially in `extras`...

File: nautobot/core/filters.py [hunk 3]
```diff
@@ -716,8 +715,23 @@ def get_filters(cls):
...
+                to_field_name="name",
+                query_params={"content_type": cls._meta.model._meta.label_lower},
+                label="Static groups (name or ID)",
+            )
+
         # django-filters has no concept of "abstract" filtersets, so we have to fake it
         if cls._meta.model is not None:
+            if "tags" in filters and isinstance(filters["tags"], TagFilter):
+                filters["tags"].extra["query_params"] = {"content_types": [cls._meta.model._meta.label_lower]}
+
```
Comment (glennmatthews) [inline nautobot/core/filters.py:734]: This is part of the logic I moved out of `get_filterset_parameter_form_field` to be more generally applicable.

File: nautobot/core/templates/generic/object_bulk_update.html [hunk 4]
```diff
@@ -29,7 +29,9 @@ <h1>{% block title %}Editing {{ table.rows|length }} {{ obj_type_plural|betterti
     <div class="row">
         <div class="col-md-8">
             <div class="panel panel-default">
-                {% render_table table 'inc/table.html' %}
+                <div class="table-responsive">
+                    {% render_table table 'inc/table.html' %}
+                </div>
```
Comment (glennmatthews) [inline nautobot/core/templates/generic/object_bulk_update.html:34]: Layout fix here - without it, in object bulk edit forms where the table of objects being edited was very wide, it would overflow its bounds instead of scrolling horizontally.

File: nautobot/core/utils/lookup.py [hunk 5]
```diff
@@ -224,8 +224,9 @@ def get_created_and_last_updated_usernames_for_model(instance):
     created_by = None
     last_updated_by = None
     try:
-        created_by_record = object_change_records.get(action=ObjectChangeActionChoices.ACTION_CREATE)
-        created_by = created_by_record.user_name
+        created_by_record = object_change_records.filter(action=ObjectChangeActionChoices.ACTION_CREATE).first()
```
Comment (glennmatthews) [inline nautobot/core/utils/lookup.py:227]: With object-change factory data present, it's possible for there to be more than one CREATE record for a given object.

File: nautobot/extras/tables.py [hunk 6]
```diff
@@ -113,6 +115,17 @@
 <a href="{% url 'extras:objectchange_list' %}?request_id={{ value }}">{{ value }}</a>
 """
 
+MEMBERS_COUNT = """
+{% load helpers %}
+{% with urlname=record.model|validated_viewname:"list" %}
+{% if urlname %}
+    <a href="{% url urlname %}?static_group={{ record.name }}">{{ record.count }}</a>
```
Comment (glennmatthews) [inline nautobot/extras/tables.py:122]:
```suggestion
    <a href="{% url urlname %}?static_groups={{ record.name }}">{{ record.count }}</a>
```

File: nautobot/extras/forms/forms.py [hunk 7]
```diff
@@ -1499,6 +1511,100 @@ class SecretsGroupFilterForm(NautobotFilterForm):
...
+            "description",
+            "tenant",
+            "tags",
+        )
+
+
+class StaticGroupFilterForm(NautobotFilterForm):
+    model = StaticGroup
+    q = forms.CharField(required=False, label="Search")
+    content_type = CSVContentTypeField(
```
Comment (HanlinMiao) [inline nautobot/extras/forms/forms.py:1542]: `CSVContentTypeField` should be renamed right? As far as I can tell, this field is only used in filter forms.

File: nautobot/extras/forms/forms.py [hunk 8]
```diff
@@ -1499,6 +1511,100 @@ class SecretsGroupFilterForm(NautobotFilterForm):
...
+            required=False,
+            query_params={"content_type": model._meta.label_lower},
+            label="Add to existing group(s)",
+        )
+        self.fields["remove_from_groups"] = DynamicModelMultipleChoiceField(
+            queryset=StaticGroup.objects.all(),
+            required=False,
+            query_params={"content_type": model._meta.label_lower},
+            label="Remove from group(s)",
+        )
```
Comment (HanlinMiao) [inline nautobot/extras/forms/forms.py:1592]: These two fields input value should be mutually exclusive

File: nautobot/extras/migrations/0107_staticgroup_staticgroupassociation.py [hunk 9]
```diff
@@ -40,15 +42,28 @@ class Migration(migrations.Migration):
                     models.ForeignKey(
                         limit_choices_to=nautobot.extras.utils.FeatureQuery("static_groups"),
                         on_delete=django.db.models.deletion.CASCADE,
-                        related_name="static_groups",
+                        related_name="static_groups_set",
```
Comment (HanlinMiao) [inline nautobot/extras/migrations/0107_staticgroup_staticgroupassociation.py:45]: Is this change intentional? Should we change this back to `static_groups`?

File: nautobot/extras/models/groups.py [hunk 10]
```diff
@@ -43,12 +63,21 @@ class StaticGroup(PrimaryModel):
     content_type = models.ForeignKey(
         to=ContentType,
         on_delete=models.CASCADE,
-        related_name="static_groups",
+        related_name="static_groups_set",
```
Comment (HanlinMiao) [inline nautobot/extras/models/groups.py:66]: Same question here

File: nautobot/extras/models/groups.py [hunk 11]
```diff
@@ -80,6 +109,11 @@ def members(self):
             pk__in=self.static_group_associations.values_list("associated_object_id", flat=True)
         )
 
+    @property
```
Comment (HanlinMiao) [inline nautobot/extras/models/groups.py:112]: Should we have a `set_members` helper function in addition to `add_members` and `remove_members`?

File: nautobot/extras/models/mixins.py [hunk 12]
```diff
@@ -180,6 +180,9 @@ class Meta:
 
     @property
     def static_groups(self):
+        """
+        Returns a QuerySet of StaticGroups that have this object as a member.
```
Comment (HanlinMiao) [inline nautobot/extras/models/mixins.py:184]:
```suggestion
        Returns a queryset of StaticGroups that have this object as a member.
```

File: nautobot/tenancy/views.py [hunk 13]
```diff
@@ -82,22 +82,23 @@ class TenantView(generic.ObjectView):
 
     def get_extra_context(self, request, instance):
         stats = {
```
Comment (HanlinMiao) [inline nautobot/tenancy/views.py:84]: Wonder if we can replace these with a for loop.

File: nautobot/extras/tests/test_views.py [hunk 14]
```diff
@@ -2766,7 +2769,48 @@ def test_list_objects_with_constrained_permission(self):
...
+        self.form_data["content_type"] = instance.content_type.pk  # Content-type is not editable after creation
+        super().test_edit_object_with_constrained_permission()
+
+
+class StaticGroupAssociationTestCase(
+    ViewTestCases.BulkDeleteObjectsViewTestCase,
+    ViewTestCases.DeleteObjectViewTestCase,
+    ViewTestCases.GetObjectViewTestCase,
+    ViewTestCases.GetObjectChangelogViewTestCase,
+    ViewTestCases.ListObjectsViewTestCase,
```
Comment (HanlinMiao) [inline nautobot/extras/tests/test_views.py:2807]: Is there any other way to access these features besides manually entering `extras/static-group-associations/` in the browser?

**Review by HanlinMiao — APPROVED** [review]

File: nautobot/core/templates/generic/object_list.html [hunk 15]
```diff
@@ -180,5 +204,27 @@ <h1>{% block title %}{{ title }}{% endblock %}</h1>
         added: jsify_form,
         removed: (row) => { row.remove(); }
     });
+
+    // TODO: move this to a JS file that is auto-included by static_group_assignment_modal templatetag
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/templates/generic/object_list.html:208]: I am mixing them, yes. The Bootstrap 3 modal functionality is jquery-dependent so I can't use only vanilla JS at this time.

File: nautobot/core/templates/utilities/templatetags/static_group_assignment_modal.html [hunk 16]
```diff
@@ -0,0 +1,37 @@
+{% load form_helpers %}
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/templates/utilities/templatetags/static_group_assignment_modal.html:1]: I considered it but felt like it would be fairly complex to implement. Value versus effort?

File: nautobot/core/templates/generic/object_list.html [hunk 17]
```diff
@@ -117,6 +118,18 @@ <h1>{% block title %}{{ title }}{% endblock %}</h1>
...
+                                            id="update_static_groups_for_all"
+                                            data-toggle="modal"
+                                            data-target="#static_group_assignment_modal"
+                                            data-objects="all"
+                                            class="btn btn-primary btn-sm"
+                                            disabled="disabled">
+                                        <span class="mdi mdi-group" aria-hidden="true"></span>
+                                        Update Static Groups for All
+                                    </button>
+                                {% endif %}
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/templates/generic/object_list.html:132]: Sorry, need what?

File: nautobot/extras/filters/__init__.py [hunk 18]
```diff
@@ -1126,6 +1130,16 @@ class StaticGroupFilterSet(NautobotFilterSet):
         },
     )
     content_type = ContentTypeMultipleChoiceFilter(choices=FeatureQuery("static_groups").get_choices, conjoined=False)
+    member_id = MultiValueUUIDFilter(
+        field_name="static_group_associations__associated_object_id",
+        label="Group member ID",
+    )
+    tenant = NaturalKeyOrPKMultipleChoiceFilter(
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/extras/filters/__init__.py:1137]: Good thought!

File: nautobot/extras/forms/forms.py (hunk 7 above)
Comment (glennmatthews) [reply nautobot/extras/forms/forms.py:1542]: Yeah, it's a misleading name at this point.

File: nautobot/extras/forms/forms.py (hunk 8 above)
Comment (glennmatthews) [reply nautobot/extras/forms/forms.py:1592]: Easier to add as form backend logic than as UI frontent logic, definitely.

File: nautobot/extras/models/groups.py (hunk 10 above)
Comment (glennmatthews) [reply nautobot/extras/models/groups.py:66]: Intentional - `.static_groups` should always be the groups that an object is the member of. We don't support static groups of ContentType at this time but in case that's ever something we add, trying to preemptively avoid a name collision here.

File: nautobot/extras/models/mixins.py (hunk 12 above)
Comment (glennmatthews) [reply nautobot/extras/models/mixins.py:184]: But it is an instance of `QuerySet` class.

File: nautobot/extras/tests/test_views.py (hunk 14 above)
Comment (glennmatthews) [reply nautobot/extras/tests/test_views.py:2807]: I *think* the changelog view on the home page might link to the detail view for a StaticGroupAssociation, but other than that, I don't believe so. Kinda similar to RelationshipAssociation in that regard.

File: nautobot/core/templates/generic/object_list.html [hunk 19]
```diff
@@ -117,6 +118,18 @@ <h1>{% block title %}{{ title }}{% endblock %}</h1>
                                 </label>
                             </div>
                             <div class="pull-right">
+                                {% if content_type.model_class.is_static_group_associable_model and perms.extras.add_staticgroupassociation %}
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/templates/generic/object_list.html:121]: IMHO that's just an extra layer of obfuscation - moving logic from the template to the view that really is relevant to the template only.

File: nautobot/core/templates/generic/object_list.html (hunk 15 above)
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/templates/generic/object_list.html:208]: I can change a couple of instances of jquery here to vanilla, at least.

File: nautobot/core/views/generic.py [hunk 20]
```diff
@@ -157,6 +157,14 @@ def get(self, request, *args, **kwargs):
                 RequestConfig(request, paginate).configure(associations_table)
                 associations_table.columns.show("pk")
                 context["associated_contacts_table"] = associations_table
+            if instance.is_static_group_associable_model:
+                paginate = {"paginator_class": EnhancedPaginator, "per_page": get_paginate_count(request)}
+                static_groups = instance.static_groups.restrict(request.user, "view")
+                static_groups_table = StaticGroupTable(static_groups, orderable=False)
+                RequestConfig(request, paginate).configure(static_groups_table)
+                static_groups_table.columns.show("pk")
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/views/generic.py:165]: Looks like no. Will change.

File: nautobot/core/utils/filtering.py [hunk 21]
```diff
@@ -114,17 +114,14 @@ def get_filterset_parameter_form_field(model, parameter, filterset=None):
     elif isinstance(field, NumberFilter):
         form_field = forms.IntegerField()
     elif isinstance(field, ModelMultipleChoiceFilter):
-        related_model = Status if isinstance(field, StatusFilter) else field.extra["queryset"].model
         form_attr = {
-            "queryset": related_model.objects.all(),
+            "queryset": field.extra["queryset"],
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/utils/filtering.py:118]: I'm thinking not, as the other two params have defaults if not present, but this one must be present.

File: nautobot/core/templates/generic/object_list.html (hunk 17 above)
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/core/templates/generic/object_list.html:132]: Sorry should made myself clearer. I am wondering if we should have what you did here for the `Add Components` button on the DeviceListView, something like `Add Components for All`
Screenshot 2024-05-06 at 3 15 08 PM

[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/templates/generic/object_list.html:132]: Ah gotcha. It'd be nice if we could come up with a way for this behavior to be automatic rather than needing to declare two different buttons for each bulk action.

File: nautobot/tenancy/views.py (hunk 13 above)
Comment (glennmatthews) [reply nautobot/tenancy/views.py:84]: Not sure I see a straightforward way to do that.

File: nautobot/core/views/renderers.py [hunk 22]
```diff
@@ -251,6 +251,13 @@ def get_context(self, data, accepted_media_type, renderer_context):
                 context["associated_contacts_table"] = associations_table
             else:
                 context["associated_contacts_table"] = None
+            if instance.is_static_group_associable_model:
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/views/renderers.py:254]: Good point, will do.

**Comment by gsnider2195** [comment]
I'm not seeing anything in the static groups detail view to help users understand where to go to add members to the group they just created. I tried clicking on the "Help" button on the static group create menu but the docs don't explain how to add members to a static group. Do we want to update the model docs or have a "user guide" docs follow-up? Perhaps if a static group is empty, on the `members` tab we can show a link to the list view for the associated content type with something like:

```html
Add members to this group from the <a href='/ipam/prefixes/'>Prefixes</a> list view.
```

**Comment by gsnider2195** [comment]
We have checkboxes on the table on the members tab but there's no bulk action buttons:

image

**Comment by glennmatthews** [comment]
> We have checkboxes on the table on the members tab but there's no bulk action buttons:

Removed the checkboxes for now in the latest commit.

**Comment by gsnider2195** [comment]
A couple of comments on the Update static groups modal, neither blocking:

- Is it possible to pop the modal up without having the page behind it scroll to the top?
- When the `Cancel` button on the modal is clicked, can the form be reset to the initial values?

File: nautobot/core/tables.py [hunk 23]
```diff
@@ -38,6 +40,14 @@ def __init__(self, *args, user=None, hide_hierarchy_ui=False, **kwargs):
         # Add custom field columns
         model = self._meta.model
 
+        if model.is_static_group_associable_model:
+            self.base_columns["static_group_count"] = LinkedCountColumn(
+                viewname="extras:staticgroup_list",
+                url_params={"member_id": "pk"},
```
Comment (gsnider2195) [inline nautobot/core/tables.py:46]: I think you brought this up in dsu. How difficult would it be to add content type to this? I know it's astronomically unlikely for the uuids to collide, so definitely not blocking

**Comment by gsnider2195** [comment]
Should the count be added to the static groups tab on the detail view?

image

File: nautobot/core/tables.py (hunk 23 above)
Comment (glennmatthews) [reply nautobot/core/tables.py:46]: A little tricky here because `LinkedCountColumn` expects each value in the `url_params` (e.g. `"pk"`) to be an instance attribute that it can `getattr()` on. So just adding `"content_type": model._meta.label_lower` or even `"content_type": "_meta.label_lower"` wouldn't work as one might expect. Might be able to make LinkedCountColumn smart enough to handle nested attributes though...

Comment (gsnider2195) [reply nautobot/core/tables.py:46]: > Might be able to make LinkedCountColumn smart enough to handle nested attributes though...

Probably not worth it for this use case

**Comment by glennmatthews** [comment]
> A couple of comments on the Update static groups modal, neither blocking:
> 
>     * Is it possible to pop the modal up without having the page behind it scroll to the top?

I don't see an option in the Bootstrap 3 modal API to change this behavior, unfortunately -  I agree it's a bit jarring.

> 
>     * When the `Cancel` button on the modal is clicked, can the form be reset to the initial values?

You mean clear the form fields? I can give that a try.

**Review by HanlinMiao — APPROVED** [review]
